README
Introduction
============

dcaenc is an open-source implementation of the DTS Coherent Acoustics codec.
Only the core part of the encoder is implemented, because the public
specification (ETSI TS 102 114 V1.2.1) is insufficient for implementing
anything else.

The project page is at http://aepatrakov.narod.ru/dcaenc/

The latest code can be obtained from git:

 git clone git://gitorious.org/dtsenc/dtsenc.git

Windows-specific notes
======================

There is a fork of dcaenc specially targetted at Windows users. You can get
it from http://gitorious.org/~mulder/dtsenc/mulders-dtsenc . As compared to
this version, that fork adds a Visual Studio project file, a progress indicator
and support for Unicode file names on Windows. It also breaks Linux support.
Still, the original dcaenc can be compiled for Windows, here is how.

The recommended compiler for dcaenc on Windows is MinGW. Get it from
http://sourceforge.net/projects/mingw/files/Installer/ . You may want to use
either the mingw-get.exe utility directly (just copy it to c:\mingw\bin,
install nothing else), or use a graphical installer. In the latter case,
install just the C compiler.

After installation, add c:\mingw\bin and c:\mingw\msys\1.0\bin to the Path
environment variable (right-click My Computer -> Properties -> Advanced ->
Environment Variables) if it is not already there. Then install the required
MinGW components from the command prompt:

 mingw-get install mingw32-base mingw32-autotools

For autotools to work, you also have to copy or rename the
c:\mingw\msys\1.0\etc\fstab.sample file to c:\mingw\msys\1.0\etc\fstab.

Installation
============

The dcaenc library has no external dependencies. The optional ALSA plugin
that comes with it, obviously, depends on the ALSA library compiled with
support for external plugins.

See the file INSTALL for the generic installation instructions.
Quick summary for UNIX-like systems:

 autoreconf -f -i -v       # only if building from git
 ./configure --prefix=/usr --libdir=??? # see below
 make
 su -c 'make install'      # Or, on Ubuntu: sudo make install

The --libdir parameter is distribution- and architecture-specific, thus, the
default (/usr/lib) may be wrong for your system. This parameter influences
the shared library and ALSA plugin location. If you specify this parameter
incorrectly, ALSA will not find the plugin. You can verify that the value is
correct by looking if there are already any ALSA plugins for the same
architecture in $libdir/alsa-lib. Here are some values to consider:

 Most 32-bit x86 linux systems: /usr/lib
 Most x86-64 linux systems: /usr/lib or /usr/lib64
 32-bit libraries on most x86-64 linux systems: /usr/lib or /usr/lib32
 x86 Debian or Ubuntu (new versions): /usr/lib/i386-linux-gnu
 x86-64 Debian or Ubuntu (new versions): /usr/lib/x86_64-linux-gnu

Quick summary for Windows:

 bash
 autoreconf -f -i -v       # only if building from git
 ./configure MAKE=mingw32-make
 mingw32-make

Package contents
================

On UNIX-like systems, the dcaenc package contains the libdcaenc.so shared
library, the corresponding dcaenc.h header, the dcaenc command line tool,
and (optionally, on Linux only) the ALSA plugin.

On Windows, you only get the command-line tool, even if you use the forked
version.

The libdcaenc library
=====================

The library may be useful if you want to create DTS streams in your own
UNIX application.

To do so, you have to call the following functions:

dcaenc_context dcaenc_create(
    int sample_rate,
    int channel_config,
    int approx_bitrate,
    int flags);

This function creates a library context according to its parameters. The
resulting context is an opaque value that should be passed to the other library
functions as the first parameter.

The sample rate must be one of the following values: 32000, 44100, 48000 or
those divided by 2 or 4. The channel configuration is one of the
DCAENC_CHANNELS_* defines listed in the dcaenc.h header. Note that values
greater than DCAENC_CHANNELS_3FRONT_2REAR always return an error now because
their encoding requires the XCH/XXCH extensions that became documented only in
September 2011 and thus were not implemented. Only non-LFE channels have to be
specified here. The approximate bitrate is specified in bits per second and
may be rounded slightly up or down by the library.

The flags parameter should be a logical OR of zero or more of the
DCAENC_FLAG_* defines. Their meanings:

DCAENC_FLAG_28BIT: use only 28 out of 32 bits in each four bytes of output.
This may be useful if the loudness of the hiss resulting from
mis-interpretation of the encoded stream as PCM must be reduced. However,
this also results in the reduction of the effective bitrate by 12.5%. DTS
CDs are usually encoded with this option.

DCAENC_FLAG_BIGENDIAN: produce a big-endian variant of the DTS bitstream.
This may be useful for writing the stream as a DVD sound track.

DCAENC_FLAG_LFE: indicates that a separate LFE channel has to be added to the
layout indicated by the channel_config parameter.

DCAENC_FLAG_PERFECT_QMF: selects a "perfect-reconstruction" version of the
quadrature mirror filter. This reduces distortions inherent in the filterbank
design, but makes the filter output more sensitive to quantization errors
introduced later in the encoder. libdca version 0.0.5 or earlier will not be
able to decode the resulting stream correctly due to a bug (mistyped table)
in it. It makes no sense to use this option for streams with bitrate of
0.3 Mbps per channel or less.

DCAENC_FLAG_IEC_WRAP: wraps DTS frames as defined by the IEC 61937-5 standard
for transmission over SPDIF. Namely, the library adds a 8-byte header to each
frame and pads the frame with zeroes to achieve the same bitrate as a stereo
16-bit PCM stream.

On any error, dcaenc_create() returns NULL. There is no way to find out the
reason for the error.

int dcaenc_channel_config_to_count(int channel_config);

Returns the number of channels used in a given channel configuration, or -1
if the channel configuration is invalid or unsupported.

int dcaenc_bitrate(dcaenc_context c);

Returns the actual bitrate that is used by the library.

int dcaenc_input_size(dcaenc_context c);

Returns the size of the input buffer that your application has to submit, in
samples. Now this value is always equal to 512.

int dcaenc_output_size(dcaenc_context c);

Returns the size of the output buffer that the application should provide,
in bytes.

int dcaenc_convert_s32(
    dcaenc_context c,
    const int32_t *input,
    uint8_t *output);

Performs the conversion of PCM samples stored in the input buffer to the
DTS bitstream, stores one frame of the encoded bitstream in the output buffer.
The input buffer should contain interleaved signed 32-bit samples. The channel
order is as follows:

DCAENC_CHANNELS_MONO: center
DCAENC_CHANNELS_STEREO: left, right
DCAENC_CHANNELS_3FRONT: center, left, right
DCAENC_CHANNELS_2FRONT_1REAR: left, right, surround
DCAENC_CHANNELS_3FRONT_1REAR: center, left, right, surround
DCAENC_CHANNELS_2FRONT_2REAR: left, right, surround left, surround right
DCAENC_CHANNELS_3FRONT_2REAR: center, left, right, surround left, surround right

If the LFE channel is used, it should be added as the last one.

The following layouts are also defined and do not return an error while
encoding, but do not result in a high-quality bitstream due to incompatibility
with the psychoacoustical model used by the library:

DCAENC_CHANNELS_DUAL_MONO: A, B
DCAENC_CHANNELS_STEREO_SUMDIFF: left+right, left-right
DCAENC_CHANNELS_STEREO_TOTAL: left total, right total

Unfortunately, there is currently no API to query the order of channels by the
channel layout. This is because it is unclear what functionality related to
channel mapping should be exposed. E.g., should speaker IDs compatible
with ksmedia.h (as used in Windows and in WAVEFORMATEXTENSIBLE wav files) be
used, even though a mapping between these speaker IDs and the positions used
in the DTS specification is in some cases only approximate (e.g., there is no
"Surround Left 2" speaker in ksmedia.h). Please mail your suggestions
regarding this API to patrakov@gmail.com.

dcaenc_convert_s32() returns the number of bytes written to the output buffer.
Right now, it is always the same as returned by dcaenc_output_size(), but
this will change if variable bitrate encoding is added to the library.

int dcaenc_destroy(dcaenc_context c, uint8_t *output);

Destroys the library context. If a non-NULL value is provided in the output
parameter, the library encodes the final frame and puts it there. This may be
useful because there is a 512-sample latency inherent in the DTS filterbank,
so the output frame gets the last portion of the PCM input submitted earlier.
The returned value indicates the number of bytes written to the output buffer.

The command-line application
============================

The dcaenc command line utility converts multichannel wav files to DTS. The
resulting DTS files can be written to an audio CD and played via a digital
connection (SPDIF or HDMI) to a receiver, or, after changing the endianness,
used as sound tracks for a DVD. Currently there are no options to select
28-bit encoding or change the endianness. This is a bug.

Usage:

 dcaenc input.wav output.dts bitrate

The input wav file should have the same channel order as defined by SMPTE,
i.e.: left, right, center, lfe, surround left, surround right.

Some destinations require a specific bitrate to be specified. To create a
CD-compatible DTS file from a multichannel file (that needs to have the sample
rate of 44100 Hz and either 16 or 32 bits per sample), run:

 dcaenc input.wav output.dts 1411200

To create a DVD-compatible track from a multichannel wav file that has the
48 kHz sample rate:

 dcaenc input.wav output.dts 1509000

or for a half-rate output:

 dcaenc input.wav output.dts 754500

and then byte-swap the resulting output.dts file. Mux it with your MPEG2 video
track using the "mplex" tool from the mjpeg-tools package.

Known bug: wav files with floating-point samples are misinterpreted as
containing 32-bit integer samples.

ALSA Plugin
===========

The ALSA plugin may be useful for playing multichannel sound from arbitrary
ALSA applications through an SPDIF link. This is needed because the SPDIF
link cannot carry enough bits per second to transport the raw uncompressed
5.1 audio.

The "alsa-plugins" package contains a similar plugin for on-the-fly AC3
encoding.

The plugin should not be used with HDMI connections, unless the video card
imposes the same audio bandwidth limitations as the SPDIF link. The HDMI
standard defines enough bandwidth so that the uncompressed 5.1 PCM stream
fits even at 192 kHz sample rate.

The ALSA plugin should work in real time on any modern CPU. Here on Intel
Core i5 @ 1.20 GHz (i.e. in powersaving mode) it eats ~40% of a single core.

To use the ALSA plugin, add the following line to your $HOME/.asoundrc
file or to /etc/asound.conf:

 <confdir:pcm/dca.conf>

It will create an additional ALSA device for each of your sound cards that
have an SPDIF output. The name of the device will be similar to
"dca:CARD=Intel,DEV=0", or, for the default card, simply "dca".

If you want to encode DTS and send it to something that is not SPDIF, add
a snippet similar to the following:

 pcm.dcahdmi {
     type dca
     slave.pcm "hdmi:CARD=Intel,DEV=0,AES0=6"
 }

The card name can be found in the "aplay -L" output, and AES0=6 parameter
marks the stream as non-audio (so that it doesn't get played by mistake --
this would result in loud hiss that can damage loudspeakers). Note, however,
that there is no known non-artificial use case for this. If you have a video
card that advertises its audio portion as the "hdmi" ALSA device, but still
has the same format limitations as SPDIF, or if you have any other reason to
define your own PCM, please send an email to patrakov@gmail.com.

Unlike the AC3 encoder, there are no other configuration parameters. This is
because it does not make sense to have them. The hard-coded settings (same
bitrate as stereo PCM, all bits used) provide the best possible quality and
should work for everyone.

To direct mplayer output to the default card via the encoder:

 mplayer -channels 6 -ao alsa:device=dca file.flac

It is not possible to use dmix on top of the encoder. This is a limitation
of dmix: it only works on direct hardware devices providing mmap, and the
dca plugin is not a hardware device and does not provide mmap. Please use
PulseAudio instead.

Known bug: the ALSA plugin doesn't report the supported sample rates correctly.
Fixing this requires rewriting the plugin from the extplug infrastructure to
ioplug. So you may need to add one of the following flags to mplayer command
line:

 -af resample=44100
 -af resample=48000

Use with PulseAudio
===================

The ALSA plugin can be used with PulseAudio. To do so, add the following lines
to the end of the /usr/share/pulseaudio/alsa-mixer/profile-sets/default.conf
file:

[Mapping iec958-dts-surround-51]
description = DTS
device-strings = dca:%f
channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe
priority = 3
direction = output

Debian and Ubuntu have a patched version of pulseaudio that uses the
/usr/share/pulseaudio/alsa-mixer/profile-sets/extra-hdmi.conf file instead
for Intel and NVidia sound cards. If that file exists, add the above lines
there, too.

Later they may be added to the default profile upstream. After restarting
PulseAudio, it will see an additional "DTS" output profile and allow you to
select it in the volume control application such as pavucontrol or
gnome-volume-control.

Compatibility
=============

The author has no DTS-capable receiver and tests the encoder only by decoding
its output with ffmpeg, libdca or ArcSoft DTS decoder (the same engine as
used in WinDVD).

The ALSA plugin has been tested and found to work with the following receivers
by other people:

 Logitech Z5500
 JVC TH-A25
 Samsung HT-Z310

Some receivers (including JVC TH-A25) mute their outputs when receiving full
32-bit DTS stream (as generated by the ALSA plugin) over SPDIF with AES0=6
(the default for the "dca" family of ALSA devices). To overcome this problem,
add the following line to the end of .asoundrc:

 defaults.pcm.dca.aes0 0x04

However, it can cause your receiver to unmute its output even if it does
not support DTS streams or does not detect them reliably. This will result
in very loud hiss that can damage the loudspeakers. So try this setting with
the lowest possible volume, and this is why it is not the default.

Similar settings exist for AES1, AES2 and AES3 SPDIF parameters.

Quality
=======

There are debates on the Internet about the relative quality of AC3 vs DTS.
AC3 uses more advanced compression algorithms, DTS allows for higher bit rates.
There were no blind tests comparing the output of the encoder with anything
else. However, this encoder uses only the most basic compression techniques
defined in the DTS specification, and thus cannot win any comparison with
commercial DTS encoders. Still, at 754 kbps, the internal psychoacoustical
model considers the distortions to be just below the threshold of detection by
human ears.

How to report bugs
==================

Bugs should be reported by email to patrakov@gmail.com, preferrably with a
short (< 10 seconds) flac sample that demonstrates the problem, and a patch
that fixes it.

Thanks
======

The following people helped me to test the encoder (including the versions that
did not work):

 Arun Raghavan
 Colin Guthrie
 Mikhail Elovskikh
 cryptonymous from the linux.org.ru forum
 rulet from the linux.org.ru forum

The following people contributed useful information:

 Adam Thomas-Murphy

The following people reported bugs:

 LoRd_MuldeR from https://gitorious.org/~mulder


NEWS
dcaenc-2: 2012-01-28
	* Minor cleanups
	* Support for 5.0 channel encoding
	* Fixed infinite loop on too-high bitrates
	* Fixed buffer overflow on attempts to encode a 7.1 file

dcaenc-1: 2011-09-17
	* Initial release.


int_data.h
/* 
 * This file is part of dcaenc.
 *
 * Copyright (c) 2008-2012 Alexander E. Patrakov <patrakov@gmail.com>
 *
 * dcaenc is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * dcaenc is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with dcaenc; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

#ifndef INT_DATA_H
#define INT_DATA_H

#include "softfloat.h"


/*
 *  Official tables taken from:
 *  ETSI TS 102 114 V1.2.1
 *
 *  Copyright (C)  European Telecommunications Standards Institute 2002.
 *  Copyright (C) European Broadcasting Union 2002.
 *  All rights reserved.
 */

/* 5.4.1 Bit Stream Header,
 * Table 5.7: RATE parameter vs. targeted bit-rate
 * with "open" replaced by 6144000
 * and "variable" and "lossless" by 16144000 (i.e. invalid value)
 */
static const int target_bitrate_table[32] = {
	32000, 56000, 64000, 96000, 112000, 128000, 192000, 224000,
	256000, 320000, 384000, 448000, 512000, 576000, 640000, 768000,
	960000, 1024000, 1152000, 1280000, 1344000, 1408000, 1411200, 1472000,
	1536000, 1920000, 2048000, 3072000, 3840000, 6144000, 16144000, 16144000,
};

/* 5.4.1 Bit Stream Header,
 * Table 5.4: Audio channel arrangement
 * FIXME: more than 5 channels are not supported
 */
static const int channels_table[16] = {
	1, 2, 2, 2, 2, 3, 3, 4, 4, 5, 6, 6, 6, 7, 8, 8,
};

/* 5.4.1 Bit Stream Header,
 * Table 5.5: Core audio sampling frequencies
 */
static const int sample_rates[9] = {
	8000, 16000, 32000, 11025, 22050, 44100, 12000, 24000, 48000,
};

static const int bitstream_sfreq[9] = {
	1, 2, 3, 6, 7, 8, 11, 12, 13,
};

/* manually derived from
 * Table B.5: Selection of quantization levels and codebooks
 * Fixme: will become invalid when Huffman codes are introduced.
 */
static const int bit_consumption[27] = {
	-8, 28, 40, 48, 52, 60, 68, 76, 80, 96,
	112, 128, 144, 160, 176, 192, 208, 224, 240, 256,
	272, 288, 304, 320, 336, 352, 368,
};

/* Table B.5: Selection of quantization levels and codebooks */
static const int quant_levels[27] = {
	1, 3, 5, 7, 9, 13, 17, 25, 32, 64,
	128, 256, 512, 1024, 2048, 4096, 8192, 16384, 32768, 65536,
	131072, 262144, 524288, 1048576, 2097152, 4194304, 8388608,
};

/* D.2.1 Lossy Quantization */
static const int32_t stepsize[27] = {
	0, 6710886, 4194304, 3355443, 2474639, 2097152, 1761608, 1426063, 796918, 461373,
	251658, 146801, 79692, 46137, 27263, 16777, 10486, 5872, 3355, 1887,
	1258, 713, 336, 168, 84, 42, 21,
};

/* Generated table, FIXME: move to math_tables.c */
/* a / stepsize[i] == mul32(a, stepsize_inv[i].m) >> stepsize_inv[i].e */
static const softfloat stepsize_inv[27] = {
	{0, 0}, {1342177360, 21}, {2147483647, 21}, {1342177360, 20},
	{1819901661, 20}, {2147483647, 20}, {1278263843, 19}, {1579032492, 19},
	{1412817763, 18}, {1220162327, 17}, {1118482133, 16}, {1917391412, 16},
	{1766017772, 15}, {1525212826, 14}, {1290553940, 13}, {2097179000, 13},
	{1677683200, 12}, {1497972244, 11}, {1310893147, 10}, {1165354136, 9},
	{1748031204, 9}, {1542092044, 8}, {1636178017, 7}, {1636178017, 6},
	{1636178017, 5}, {1636178017, 4}, {1636178017, 3},
};

/* D.1.2 7-bit Quantization (Nominal 1.1 dB Step) */
static const int32_t scalefactor[128] = {
	1, 1, 2, 2, 2, 2, 3, 3,
	3, 4, 4, 5, 6, 7, 7, 8,
	10, 11, 12, 14, 16, 18, 20, 23,
	26, 30, 34, 38, 44, 50, 56, 64,
	72, 82, 93, 106, 120, 136, 155, 176,
	200, 226, 257, 292, 331, 376, 427, 484,
	550, 624, 708, 804, 912, 1035, 1175, 1334,
	1514, 1718, 1950, 2213, 2512, 2851, 3236, 3673,
	4169, 4732, 5370, 6095, 6918, 7852, 8913, 10116,
	11482, 13032, 14791, 16788, 19055, 21627, 24547, 27861,
	31623, 35892, 40738, 46238, 52481, 59566, 67608, 76736,
	87096, 98855, 112202, 127350, 144544, 164059, 186209, 211349,
	239883, 272270, 309030, 350752, 398107, 451856, 512861, 582103,
	660693, 749894, 851138, 966051, 1096478, 1244515, 1412538, 1603245,
	1819701, 2065380, 2344229, 2660725, 3019952, 3427678, 3890451, 4415704,
	5011872, 5688529, 6456542, 7328245, 8317638, 8317638, 8317638, 8317638,
};

/* Generated table */
/* sqrt(2) * a / scalefactor[i] == 4 * mul32(a, scalefactor_inv[i].m) >> scalefactor_inv[i].e */
static const softfloat scalefactor_inv[128] = {
	{2147483647, 1}, {2147483647, 1}, {2147483647, 2}, {2147483647, 2},
	{2147483647, 2}, {2147483647, 2}, {1431655765, 2}, {1431655765, 2},
	{1431655765, 2}, {2147483647, 3}, {2147483647, 3}, {1717986918, 3},
	{1431655765, 3}, {1227133513, 3}, {1227133513, 3}, {2147483647, 4},
	{1717986918, 4}, {1561806289, 4}, {1431655765, 4}, {1227133513, 4},
	{2147483647, 5}, {1908874353, 5}, {1717986918, 5}, {1493901668, 5},
	{1321528398, 5}, {1145324612, 5}, {2021161080, 6}, {1808407282, 6},
	{1561806289, 6}, {1374389534, 6}, {1227133513, 6}, {2147483647, 7},
	{1908874353, 7}, {1676084798, 7}, {1477838209, 7}, {1296593900, 7},
	{1145324612, 7}, {2021161080, 8}, {1773405851, 8}, {1561806289, 8},
	{1374389534, 8}, {1216273924, 8}, {2139127680, 9}, {1882725390, 9},
	{1660893697, 9}, {1462116526, 9}, {1287484341, 9}, {1135859119, 9},
	{1999112050, 10}, {1762037865, 10}, {1552982525, 10}, {1367551775, 10},
	{1205604855, 10}, {2124660150, 11}, {1871509153, 11}, {1648443220, 11},
	{1452459217, 11}, {1279990253, 11}, {1127704233, 11}, {1987368509, 12},
	{1750814693, 12}, {1542632939, 12}, {1359099663, 12}, {1197398995, 12},
	{2109880792, 13}, {1858853132, 13}, {1638006149, 13}, {1443165385, 13},
	{1271479187, 13}, {1120235993, 13}, {1973767086, 14}, {1739045674, 14},
	{1532153461, 14}, {1349922194, 14}, {1189384493, 14}, {2095804865, 15},
	{1846464029, 15}, {1626872524, 15}, {1433347133, 15}, {1262853884, 15},
	{1112619678, 15}, {1960569045, 16}, {1727349015, 16}, {1521881227, 16},
	{1340842289, 16}, {1181357555, 16}, {2081669156, 17}, {1834047752, 17},
	{1615889229, 17}, {1423675973, 17}, {1254322457, 17}, {1105123583, 17},
	{1947330755, 18}, {1715693602, 18}, {1511607799, 18}, {1331801790, 18},
	{1173384427, 18}, {2067616532, 19}, {1821667648, 19}, {1604980024, 19},
	{1414066955, 19}, {1245861410, 19}, {1097665748, 19}, {1934193616, 20},
	{1704119624, 20}, {1501412075, 20}, {1322817107, 20}, {1165466323, 20},
	{2053666205, 21}, {1809379407, 21}, {1594151671, 21}, {1404526328, 21},
	{1237455941, 21}, {1090259329, 21}, {1921143210, 22}, {1692621231, 22},
	{1491281857, 22}, {1313892269, 22}, {1157603482, 22}, {2039810470, 23},
	{1797172644, 23}, {1583396912, 23}, {1395050052, 23}, {1229107276, 23},
	{1082903494, 23}, {1082903494, 23}, {1082903494, 23}, {1082903494, 23},
};
#endif
